File: solutions/go/say/1/say.go
package say

import "fmt"

const (
	zero = "zero"
)

var unitaries = []string{zero, "one", "two", "three", "four", "five", "six", "seven",
	"eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen",
	" sixteen", "seventeen", "eighteen", "nineteen"}

var tens = []string{"", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"}

var units = []struct {
	limit   int64
	divisor int64
	name    string
}{
	{1_000, 100, "hundred"}, {1_000_000, 1_000, "thousand"}, {1_000_000_000, 1_000_000, "million"},
	{1_000_000_000_000, 1_000_000_000, "billion"}}

func Say(n int64) (string, bool) {
	if n < 0 {
		return "", false

	}
	if n < 20 {
		return unitaries[n], true
	}
	if n < 100 {
		unitary := unitaries[n%10]
		ten := tens[n/10]
		if unitary == zero {
			return ten, true
		} else {
			return fmt.Sprintf("%s-%s", ten, unitary), true
		}
	}
	for _, unit := range units {
		if n >= unit.limit {
			continue
		}
		bottom, _ := Say(n % unit.divisor)
		top, _ := Say(n / unit.divisor)
		if bottom == zero {
			return fmt.Sprintf("%s %s", top, unit.name), true
		} else {
			return fmt.Sprintf("%s %s %s", top, unit.name, bottom), true
		}
	}
	return "", false
}

File: solutions/javascript/say/1/say.js

const NAMES = {0: 'zero', 1: 'one', 2: 'two', 3: 'three', 4: 'four', 5: 'five', 6: 'six',
              7: 'seven', 8: 'eight', 9: 'nine', 10: 'ten', 11: 'eleven', 12: 'twelve', 13: 'thirteen',
              14: 'fourteen', 15: 'fifteen', 16: 'sixteen', 17: 'seventeen', 18: 'eighteen', 19: 'nineteen',
              20: 'twenty', 30: 'thirty', 40: 'forty', 50: 'fifty', 60: 'sixty', 70: 'seventy', 80: 'eighty',
              90: 'ninety'}

export function say(n) {

  if (n < 0 || n > 999999999999) throw new Error('Number must be between 0 and 999,999,999,999.')
  
  if (n <= 20) return NAMES[n]

  if (n < 100) {
    const d = Math.trunc(n / 10) * 10
    const r = n - d
    return (r == 0) ? NAMES[d] : `${NAMES[d]}-${say(r)}`
  }
  
  if (n < 1000) {
    const d = Math.trunc(n / 100)
    const r = n - d * 100
    return (r == 0) ? `${NAMES[d]} hundred` : `${NAMES[d]} hundred ${say(r)}`
  }

  if (n < 1000000) {
    const d = Math.trunc(n / 1000)
    const r = n - d * 1000
    return (r == 0) ? `${say(d)} thousand` : `${say(d)} thousand ${say(r)}` 
  }

  if (n < 1000000000) {
    const d = Math.trunc(n / 1000000)
    const r = n - d * 1000000
    return (r == 0) ? `${say(d)} million` : `${say(d)} million ${say(r)}` 
  }

  const d = Math.trunc(n / 1000000000)
  const r = n - d * 1000000000
  return (r == 0) ? `${say(d)} billion` : `${say(d)} billion ${say(r)}` 
};
